### ISSUE TITLE
Updating Django's SECRET_KEY makes encrypted fields inaccessible

### ISSUE THREAD

**Issue by hendrikschneider** [frgmt/django-fernet-encrypted-fields#13]
I've just tested rotating Django's SECRET_KEY and my field's couldn't be encrypted anymore even though a SALT_KEY is set:

How to replicate:
1) Create a new django project
2) Create a new app with an EncryptedTextField
3) Start Django shell
4) Create an instance with some example text and save
5) Stop Django shell
6) Update secret key
7) Load earlier data from db and print content of the text field  
```
# Settings
# orignal key
SECRET_KEY = 'django-insecure-i40*4wu4erqc24em97k&fy#bj^rg%25)mtf2_cj)=2u7eg(pht'

# rotated key
# SECRET_KEY = 'django-insecure-i40*4wu4erqc24em97k&fy#bj^rg%25)mtf2_cj)=2u7eg(ph3'

SALT_KEY = '0123456789abcdefghijklmnopqrstuvwxyz'
```

Model
```
from encrypted_fields.fields import EncryptedTextField

from django.db import models


class MyModel(models.Model):
    text_field = EncryptedTextField()
```

```
Console output
In [3]: MyModel.objects.create(text_field="test1")
Out[3]: <MyModel: MyModel object (1)>

In [4]: m = MyModel.objects.last()

In [5]: m.text_field
Out[5]: 'test1'

###########
# Rotate key #
###########

In [2]: from test_app.models import MyModel

In [3]: m = MyModel.objects.last()

In [4]: m.text_field
Out[4]: 'gAAAAABj8is-Hi0WTGpbcC5Xg1J8CYO8Zh5ErsWW631UZ2m3LimZkRrCZrII83VJAy831W9ISWCja1SQecwAA7eZ0panQuNY4g==
```

**Comment by hendrikschneider** [comment]
I've also tried to add the old secret key to the list of SALT_KEY but it also didn't change the outcome.

`SALT_KEY = ['0123456789abcdefghijklmnopqrstuvwxyz', "django-insecure-i40*4wu4erqc24em97k&fy#bj^rg%25)mtf2_cj)=2u7eg(pht"]
`

@frgmt Do you have an idea, why it is like that? When I checked the code, I didn't see anything obvious that would cause that behavior.

**Comment by naohide** [comment]
@StevenMapes Do you know why this happens?

**Comment by StevenMapes** [comment]
> @StevenMapes Do you know why this happens?

For the reason @hendrikschneider has said, it's used in the function. I hadn't noticed it before and I'd agree about changing it to use another value first and then falling back to the secret key that way it could be rotated much like how I changed this project to allow for salt rotation. What we may want to do though I'd to limit it to a maximum of two values effectively old and new as otherwise if you had 6 salts and 6 keys it would produce 36 potential attempts to decrypt which runs the risk of becoming slow.  Or perhaps leaving or open and having a warning on the docs.

**Comment by hendrikschneider** [comment]
@StevenMapes @naohide I think limiting is a good idea. Actually, only two keys are needed to support key rotation and it would offer a clear way to document how to rotate the encryption secret. Having this approach would lead people to have only key active during normal operations and two in the moment they want to rotate the key and afterwards again only one key. No way to get lazy while rotating.

**Comment by StevenMapes** [comment]
I like the idea of the management command to check and rotate values.

Using DJANGO_FERNET_FIELD_ENCRYPTION_KEYS as a list makes sense to me as, whilst long, it's explicit and verbose.

**Comment by blag** [comment]
As of Django 4.1, Django use a [`SECRET_KEY_FALLBACKS` setting](https://docs.djangoproject.com/en/4.2/ref/settings/#secret-key-fallbacks), which is used to rotate secret keys. If users use that properly, this project should be able to use that instead of `DJANGO_FERNET_FIELD_ENCRYPTION_KEYS`.

**Comment by flaiming** [comment]
> As of Django 4.1, Django use a [`SECRET_KEY_FALLBACKS` setting](https://docs.djangoproject.com/en/4.2/ref/settings/#secret-key-fallbacks), which is used to rotate secret keys. If users use that properly, this project should be able to use that instead of `DJANGO_FERNET_FIELD_ENCRYPTION_KEYS`.

This is actually not true, because django-fernet-encrypted-fields is using just `SECRET_KEY`, not `SECRET_KEY_FALLBACKS`. So when I change the `SECRET_KEY`, and set the old one to `SECRET_KEY_FALLBACKS`, I still cannot use the old encrypted fields and resave them, I'm just getting encrypted data. https://github.com/jazzband/django-fernet-encrypted-fields/blob/adc35e961d973246d3d28e4cd8671277d37894a9/encrypted_fields/fields.py#L42

**Comment by flaiming** [comment]
I just created a #24 for this.

**Comment by rmcsharry** [comment]
I think this issue can be closed, since the linked PR fixed it, right?

**Comment by StevenMapes** [comment]
Agreed. The change to support this has been implemented by supporting the SECRET_KEY_FALLBACKS on line 31

`secret_keys = [settings.SECRET_KEY] + getattr(settings, "SECRET_KEY_FALLBACKS", list())`
